### ISSUE TITLE
cast(date as varchar) returns incorrect results

### ISSUE THREAD

**Issue by mbasmanova** [facebookincubator/velox#5907]
### Bug description

Cast from DATE to VARCHAR uses DATE()->toString(), which uses gmtime. 

When query runs on multiple threads, DATE()->toString() returns incorrect results because gmtime is not thread-safe.

https://cplusplus.com/reference/ctime/gmtime/

```
The returned value points to an internal object whose validity or value may be altered by any subsequent call to gmtime or [localtime](https://cplusplus.com/localtime).
```

A fix would be to switch to gmtime_r.

https://linux.die.net/man/3/gmtime_r

Repro:

```
TEST_F(MergeJoinTest, yyy) {
  std::vector<int32_t> days = {19492, 19493, 19494, 19495};
  std::vector<std::string> dates;
  for (auto d : days) {
    dates.push_back(DATE()->toString(d));
  }

  auto t0 = std::thread([&]() {
    for (auto i = 0; i < 100'000; ++i) {
      VELOX_CHECK_EQ(DATE()->toString(days[0]), dates[0], "t0: {}", i);
    }
  });

  auto t1 = std::thread([&]() {
    for (auto i = 0; i < 100'000; ++i) {
      VELOX_CHECK_EQ(DATE()->toString(days[1]), dates[1], "t1: {}", i);
    }
  });

  t0.join();
  t1.join();
}
```

CC: @aditi-pandit @kagamiori @laithsakka @vermapratyush @amitkdutta @zacw7 

### System information

n/a

**Comment by JasmineWeii** [comment]
I will work on the fix in a few days, love the investigation and repro!

**Comment by aditi-pandit** [comment]
@mbasmanova , @laithsakka : 
Someone had previously worked on https://github.com/facebookincubator/velox/pull/5029 which didn't get merged.

**Comment by mbasmanova** [comment]
There is a duplicate issue https://github.com/facebookincubator/velox/issues/5028

**Comment by mbasmanova** [comment]
Fixed in #5029
